src/fluid/braginskii/bragThermalDiffusion.hpp

```diff
@@ -0,0 +1,676 @@
+// ***********************************************************************************
+// Idefix MHD astrophysical code
+// Copyright(C) Geoffroy R. J. Lesur <geoffroy.lesur@univ-grenoble-alpes.fr>
+// and other code contributors
+// Licensed under CeCILL 2.1 License, see COPYING for more information
+// ***********************************************************************************
+
+#ifndef FLUID_BRAGINSKII_BRAGTHERMALDIFFUSION_HPP_
+#define FLUID_BRAGINSKII_BRAGTHERMALDIFFUSION_HPP_
+
+#include <string>
+
+#include "idefix.hpp"
+#include "input.hpp"
+#include "grid.hpp"
+#include "fluid_defs.hpp"
+#include "eos.hpp"
+#include "bragLimiter.hpp"
+
+// Forward class hydro declaration
+template <typename Phys> class Fluid;
+
+class DataBlock;
+
+class BragThermalDiffusion {
+ public:
+  template <typename Phys>
+  BragThermalDiffusion(Input &, Grid &, Fluid<Phys> *);  // Initialisation
+
+  void ShowConfig(); // display configuration
+
+  void AddBragDiffusiveFlux(int, const real, const IdefixArray4D<real> &);
+
+  template<const Limiter>
+  void AddBragDiffusiveFluxLim(int, const real, const IdefixArray4D<real> &);
+
+  // Enroll user-defined thermal conductivity
+  void EnrollBragThermalDiffusivity(BragDiffusivityFunc);
+
+  IdefixArray3D<real> heatSrc;  // Source terms of the thermal operator
+  IdefixArray3D<real> knorArr;
+  IdefixArray3D<real> kparArr;
+
+  // pre-computed geometrical factors in non-cartesian geometry
+  IdefixArray1D<real> one_dmu;
+
+ private:
+  DataBlock *data;
+
+  // status of the module
+  ParabolicModuleStatus &status;
+
+  BragDiffusivityFunc diffusivityFunc;
+
+  bool haveSlopeLimiter;
+
+  // helper array
+  IdefixArray4D<real> &Vc;
+  IdefixArray4D<real> &Vs;
+  IdefixArray3D<real> &dMax;
+
+  // constant diffusion coefficient (when needed)
+  real knor, kpar;
+
+  // equation of state (required to get the heat capacity)
+  EquationOfState *eos;
+
+  Limiter limiter;
+};
+
+#include "fluid.hpp"
+#include "dataBlock.hpp"
```

Comment:
To be completely honest, I didn't ask myself all these questions because I just copy-pasted the structure of the thermalDiffusion.hpp file and adapted it to my case. So I would shamelessly return the question to @glesur about the structure of the initial file...

Reply:
It's to avoid circular dependency in the declaration of the class BragThermalDiffusion. For that declaration, we make a forward declaration of Fluid (so that a Fluid can use BragThermalDiffusion without trouble), but to *define* the class constructor, we need to know what a Fluid is made off, so the include comes in right before the constructor definition, but *after* the class declaration (yeah, c++ is funny sometimes).

Same hack is used in many classes that depends on fluid, actually.


So, it is logical.

Reply:
Ok thanks geoffroy for the explanation, I close this comment then.